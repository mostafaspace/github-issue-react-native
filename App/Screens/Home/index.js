import React from 'react';
import {View, Text, Image, ScrollView} from 'react-native';
import {Actions} from 'react-native-router-flux';
import {Issues} from '../../Controllers';
import {Input} from '../../Components/Input';
import {Button} from '../../Components/Button';
import {ErrorModal} from '../../Components/ErrorModal';
import {useFormValidation} from '../../Helpers';
import {Images} from '../../Resources';
import validations from '../../Helpers/validations';
import {useDispatch} from 'react-redux';
import {POPULATE_LIST} from '../../Actions/IssuesActions';
import Styles from './styles';

const initState = {
    Organization: '',
    Repo: '',
    Page: 1,
};
export default function HomeScreen() {
    const dispatch = useDispatch();
    const {values, errors, handleChange, handleSubmit, isSubmitting, setSubmitting, showModal, modalValues} = useFormValidation(
        initState,
        validations,
        search,
    );

    async function search() {
        Issues(values, showModal, setSubmitting).then((response) => {
            if (response) {
                dispatch({type: POPULATE_LIST, payload: {response, values}});
                Actions.IssuesResult();
            }
        });
    }

    return (
        <ScrollView>
            <View style={Styles.container}>
                <Image source={Images.apc} style={Styles.apcImage} />
                <Input
                    label={'Organization'}
                    iconName={'md-logo-github'}
                    error={errors.Organization}
                    value={values.Organization}
                    maxLength={6}
                    onChangeText={(value) => handleChange('Organization', value)}
                    testID={'Org'}
                />
                <Input
                    label={'Repository'}
                    iconName={'md-git-merge-sharp'}
                    error={errors.Repo}
                    value={values.Repo}
                    maxLength={6}
                    onChangeText={(value) => handleChange('Repo', value)}
                    testID={'Repo'}
                />
                <Button onPress={handleSubmit} isLoading={isSubmitting} text={'Find'} testID={'Submit'} />
                <ErrorModal isVisible={modalValues.show} title={modalValues.text} hide={() => showModal(false, '')} testID={'ErrorModal'} />
            </View>
        </ScrollView>
    );
}

HomeScreen.navigationOptions = () => {
    return {
        headerTitle: () => <Text style={Styles.headerTitle}>{'GitHub Issues'}</Text>,
    };
};
