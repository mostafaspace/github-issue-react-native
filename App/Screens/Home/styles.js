import {StyleSheet} from 'react-native';

export default StyleSheet.create({
    headerTitle: {
        fontWeight: 'bold',
        fontSize: 16,
    },
    container: {
        flex: 1,
        justifyContent: 'center',
        alignItems: 'center',
        paddingVertical: 32,
    },
    apcImage: {
        width: 160,
        height: 160,
        marginVertical: 60,
        resizeMode: 'contain',
    },
});
