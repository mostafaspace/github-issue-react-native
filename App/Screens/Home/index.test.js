import React from 'react';
import {render, cleanup, fireEvent} from '@testing-library/react-native';
import Home from '../Home';
import renderer from 'react-test-renderer';
import '@testing-library/jest-native/extend-expect';

afterEach(cleanup);
it('renders component', () => {
    render(<Home />);
});

test('check button text', () => {
    const {getByTestId} = render(<Home />);

    //check if button have the correct text
    expect(getByTestId('Submit')).toHaveTextContent('Find');
});

test('test validations', () => {
    const {getByTestId, queryByText} = render(<Home />);

    //check if both inputs are empty
    fireEvent(getByTestId('Org'), 'onChangeText', '');
    fireEvent(getByTestId('Repo'), 'onChangeText', '');
    fireEvent.press(getByTestId('Submit'));
    expect(queryByText('Organization is Required')).toBeDefined();
    expect(queryByText('Repository is Required')).toBeDefined();

    //check if one input is empty
    fireEvent(getByTestId('Org'), 'onChangeText', 'Microsoft');
    fireEvent(getByTestId('Repo'), 'onChangeText', '');
    fireEvent.press(getByTestId('Submit'));
    expect(queryByText('Organization is Required')).toBeNull();
    expect(queryByText('Repository is Required')).toBeDefined();

    //check if one input is empty
    fireEvent(getByTestId('Org'), 'onChangeText', '');
    fireEvent(getByTestId('Repo'), 'onChangeText', 'typescript');
    fireEvent.press(getByTestId('Submit'));
    expect(queryByText('Organization is Required')).toBeDefined();
    expect(queryByText('Repository is Required')).toBeNull();

    //check if both input are full and modal is shown
    fireEvent(getByTestId('Org'), 'onChangeText', 'Microsoft');
    fireEvent(getByTestId('Repo'), 'onChangeText', 'typescript');
    fireEvent.press(getByTestId('Submit'));
    expect(queryByText('Organization is Required')).toBeNull();
    expect(queryByText('Repository is Required')).toBeNull();
});

it('Home matches snapshot', () => {
    const tree = renderer.create(<Home />).toJSON();
    expect(tree).toMatchSnapshot();
});
