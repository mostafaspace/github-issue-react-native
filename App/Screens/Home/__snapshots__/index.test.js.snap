// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home matches snapshot 1`] = `
<RCTScrollView>
  <View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
          "paddingVertical": 32,
        }
      }
    >
      <Image
        source={
          Object {
            "testUri": "../../../App/Resources/Images/assets/apc.png",
          }
        }
        style={
          Object {
            "height": 160,
            "marginVertical": 60,
            "resizeMode": "contain",
            "width": 160,
          }
        }
      />
      <View
        style={
          Object {
            "paddingHorizontal": 16,
            "width": "100%",
          }
        }
        testID="Org"
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "white",
              "borderBottomColor": "#e7e8ea",
              "borderBottomWidth": 1,
              "borderRadius": 16,
              "elevation": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
              "paddingHorizontal": 16,
              "paddingVertical": 4,
              "shadowColor": "#000",
              "shadowOffset": Object {
                "height": 1,
                "width": 0,
              },
              "shadowOpacity": 0.18,
              "shadowRadius": 1,
            }
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "#575757",
                  "fontSize": 28,
                },
                Object {
                  "height": 30,
                  "marginEnd": 4,
                  "marginVertical": 8,
                  "width": 30,
                },
                Object {
                  "fontFamily": "Ionicons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            
          </Text>
          <TextInput
            allowFontScaling={true}
            onChangeText={[Function]}
            placeholder="Organization"
            rejectResponderTermination={true}
            style={
              Object {
                "color": "#222",
                "flex": 1,
                "paddingVertical": 16,
                "textAlign": "left",
              }
            }
            testID="Input"
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
        <Text
          style={
            Object {
              "alignSelf": "flex-start",
              "color": "red",
              "fontSize": 10,
              "fontWeight": "bold",
              "lineHeight": 14,
              "marginBottom": 4,
              "marginHorizontal": 8,
              "paddingTop": 1,
              "textAlign": "left",
            }
          }
        />
      </View>
      <View
        style={
          Object {
            "paddingHorizontal": 16,
            "width": "100%",
          }
        }
        testID="Repo"
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "white",
              "borderBottomColor": "#e7e8ea",
              "borderBottomWidth": 1,
              "borderRadius": 16,
              "elevation": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
              "paddingHorizontal": 16,
              "paddingVertical": 4,
              "shadowColor": "#000",
              "shadowOffset": Object {
                "height": 1,
                "width": 0,
              },
              "shadowOpacity": 0.18,
              "shadowRadius": 1,
            }
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "#575757",
                  "fontSize": 28,
                },
                Object {
                  "height": 30,
                  "marginEnd": 4,
                  "marginVertical": 8,
                  "width": 30,
                },
                Object {
                  "fontFamily": "Ionicons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            
          </Text>
          <TextInput
            allowFontScaling={true}
            onChangeText={[Function]}
            placeholder="Repository"
            rejectResponderTermination={true}
            style={
              Object {
                "color": "#222",
                "flex": 1,
                "paddingVertical": 16,
                "textAlign": "left",
              }
            }
            testID="Input"
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
        <Text
          style={
            Object {
              "alignSelf": "flex-start",
              "color": "red",
              "fontSize": 10,
              "fontWeight": "bold",
              "lineHeight": 14,
              "marginBottom": 4,
              "marginHorizontal": 8,
              "paddingTop": 1,
              "textAlign": "left",
            }
          }
        />
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#149ADE",
            "borderRadius": 16,
            "elevation": 1,
            "flexDirection": "row",
            "justifyContent": "space-evenly",
            "margin": 16,
            "opacity": 1,
            "paddingHorizontal": 16,
            "paddingVertical": 16,
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 1,
              "width": 0,
            },
            "shadowOpacity": 0.18,
            "shadowRadius": 1,
          }
        }
        testID="Submit"
      >
        <Text
          numberOfLines={1}
          style={
            Object {
              "alignSelf": "center",
              "color": "#ffffff",
              "flex": 1,
              "paddingHorizontal": 16,
              "textAlign": "center",
            }
          }
          testID="ButtonText"
        >
          Find
        </Text>
      </View>
      <Modal
        animationType="fade"
        hardwareAccelerated={false}
        onRequestClose={[Function]}
        testID="ErrorModal"
        transparent={true}
        visible={false}
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignContent": "center",
              "alignItems": "center",
              "backgroundColor": "rgba(20, 154, 222,0.8)",
              "flex": 1,
              "justifyContent": "center",
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "white",
                "borderRadius": 30,
                "flexDirection": "column",
                "height": null,
                "justifyContent": "space-between",
                "width": "85%",
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "borderTopLeftRadius": 30,
                  "borderTopRightRadius": 30,
                  "padding": 8,
                  "width": "100%",
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#575757",
                    "fontSize": 10,
                    "fontWeight": "bold",
                  }
                }
              >
                Error
              </Text>
            </View>
            <View
              style={
                Object {
                  "alignItems": "center",
                  "margin": 16,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#575757",
                    "fontSize": 16,
                    "fontWeight": "bold",
                    "lineHeight": 24,
                    "marginHorizontal": 8,
                    "textAlign": "center",
                  }
                }
              >
                Something Went Wrong
              </Text>
            </View>
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#dddede",
                  "borderBottomLeftRadius": 30,
                  "borderBottomRightRadius": 30,
                  "borderRadius": 16,
                  "borderTopLeftRadius": 0,
                  "borderTopRightRadius": 0,
                  "elevation": 1,
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "margin": 0,
                  "opacity": 1,
                  "paddingHorizontal": 16,
                  "paddingVertical": 16,
                  "shadowColor": "#000",
                  "shadowOffset": Object {
                    "height": 1,
                    "width": 0,
                  },
                  "shadowOpacity": 0.18,
                  "shadowRadius": 1,
                  "width": "100%",
                }
              }
              testID="Button"
            >
              <Text
                numberOfLines={1}
                style={
                  Object {
                    "alignSelf": "center",
                    "color": "#ffffff",
                    "flex": 1,
                    "paddingHorizontal": 16,
                    "textAlign": "center",
                  }
                }
                testID="ButtonText"
              >
                Ok
              </Text>
            </View>
          </View>
        </View>
      </Modal>
    </View>
  </View>
</RCTScrollView>
`;
